File: src/main/java/me/pljr/pljrapispigot/listeners/BungeeListeners.java
package me.pljr.pljrapispigot.listeners;

import com.google.common.io.ByteArrayDataInput;
import com.google.common.io.ByteStreams;
import me.pljr.pljrapispigot.utils.ChatUtil;
import me.pljr.pljrapispigot.utils.PlayerUtil;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.messaging.PluginMessageListener;
import org.jetbrains.annotations.NotNull;

public class BungeeListeners implements PluginMessageListener {

    @Override
    public void onPluginMessageReceived(String channel, @NotNull Player player, byte[] message) {
        if (channel.equals("pljrapi:chat")){
            ByteArrayDataInput in = ByteStreams.newDataInput(message);
            String subChannel = in.readUTF();
            if (subChannel.equals("message")){
                receiveMessage(in.readUTF(), in.readUTF());
                return;
            }
            if (subChannel.equals("broadcast")){
                receiveBroadcast(in.readUTF(), in.readUTF());
            }
        }
    }

    private void receiveMessage(String receiver, String message){
        if (PlayerUtil.isPlayer(receiver)){
            Player player = Bukkit.getPlayer(receiver);
            ChatUtil.sendMessage(player, message);
        }
    }

    private void receiveBroadcast(String message, String perm){
        ChatUtil.broadcast(message, perm, false);
    }
}

File: src/main/kotlin/me/pljr/pljrapispigot/bungeecord/BungeeListeners.kt
package me.pljr.pljrapispigot.bungeecord

import com.google.common.io.ByteStreams
import me.pljr.pljrapispigot.player.isPlayer
import me.pljr.pljrapispigot.util.broadcastMessage
import me.pljr.pljrapispigot.util.sendMessage
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.messaging.PluginMessageListener

const val PLJRAPI_BUNGEE_CHANNEL = "pljrapi:chat"

class BungeeListeners : PluginMessageListener {

    override fun onPluginMessageReceived(channel: String, player: Player, message: ByteArray) {
        if (channel == PLJRAPI_BUNGEE_CHANNEL) {
            val input = ByteStreams.newDataInput(message)
            val subChannel = input.readUTF()
            when (subChannel) {
                "message" -> {
                    val playerName = input.readUTF()
                    val message = input.readUTF()
                    if (isPlayer(playerName)){
                        val player = Bukkit.getPlayer(playerName)!!
                        sendMessage(player, message)
                    }
                }
                "broadcast" -> {
                    val message = input.readUTF()
                    val perm = input.readUTF()
                    broadcastMessage(message, perm)
                }
            }
        }
    }
}